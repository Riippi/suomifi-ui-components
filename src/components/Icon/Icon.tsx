import React, { Component } from 'react';
import { default as styled } from 'styled-components';
import { Svg, SvgProps } from '../Svg/Svg';
import classnames from 'classnames';

export interface IconProps extends SvgProps {}

const baseClassName = 'fi-icon';

const StyledSvg = styled((props: SvgProps) => <Svg {...props} />)`
  display: inline-block;
  width: 24px;
  height: 24px;
`;

export class Icon extends Component<IconProps> {
  render() {
    const { className, ...passProps } = this.props;
    return (
      <StyledSvg
        {...passProps}
        className={classnames(className, baseClassName)}
      />
    );
  }
}
