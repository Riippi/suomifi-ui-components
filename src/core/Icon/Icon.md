```jsx
import { Icon } from 'suomifi-ui-components';

<Icon ariaLabel="Login here" className="my-icon--test" />;
```

### Icon with no label

```jsx
import { Icon } from 'suomifi-ui-components';

<Icon icon="heart" />;
```

```jsx noeditor
import { Icon } from 'suomifi-ui-components';
import { default as styled } from 'styled-components';
import { allIcons, allStaticIcons } from 'suomifi-icons';
import clipboardCopy from 'clipboard-copy';
const StyledIcon = styled(props => <Icon {...props} />)({
  height: '50px',
  width: 'auto',
  margin: '8px'
});

<div>
  <div>
    {allIcons.map(icon => (
      <StyledIcon
        mousePointer
        icon={icon}
        key={icon}
        onClick={() => clipboardCopy(icon)}
      />
    ))}
  </div>
  <div>
    {allStaticIcons.map(icon => (
      <StyledIcon
        mousePointer
        icon={icon}
        key={icon}
        onClick={() => clipboardCopy(icon)}
      />
    ))}
  </div>
</div>;
```
