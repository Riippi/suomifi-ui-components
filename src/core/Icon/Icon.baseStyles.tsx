import { css } from 'styled-components';
import { SuomifiIconProps } from './Icon';

export const iconBaseStyles = ({
  mousePointer = false,
}: SuomifiIconProps) => css`
  display: inline-block;
  vertical-align: baseline;
  ${!!mousePointer &&
    `cursor: pointer;
      & * {
        cursor: inherit;
      }
  `}
`;
