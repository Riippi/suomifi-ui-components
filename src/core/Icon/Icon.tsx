import React, { Component } from 'react';
import { default as styled } from 'styled-components';
import { iconBaseStyles } from './Icon.baseStyles';
import { withSuomifiDefaultProps } from '../theme/utils';
import { TokensProp } from '../theme';
import {
  ariaLabelOrHidden,
  ariaFocusableNoLabel,
} from '../../components/utils/aria';
import { SuomifiIcon, SuomifiIconInterface } from 'suomifi-icons';
import { logger } from '../../utils/logger';
export { BaseIconKeys } from 'suomifi-icons';

export interface IconBaseProps extends TokensProp {
  /** Aria-label for SVG, undefined hides SVG from screen reader
   * @default undefined
   */
  ariaLabel?: string;
  /** Show mouse cursor as hand-pointer */
  mousePointer?: boolean;
  testId?: string;
}

export interface IconProps extends IconBaseProps, SuomifiIconInterface {}

export const iconLogger = (
  ariaLabel: string | undefined,
  className: string | undefined,
) => {
  logger.warn(
    `Icon ERROR${
      !!ariaLabel
        ? ` with aria-label: ${ariaLabel}`
        : !!className
        ? ` with className: ${className}`
        : ''
    }`,
  );
};

/**
 * Apply Suomifi styles to Icon
 */
const StyledSuomifiIcon = styled(
  ({ ariaLabel, mousePointer, ...passProps }: IconProps) => {
    return (
      <SuomifiIcon
        {...passProps}
        {...ariaLabelOrHidden(ariaLabel)}
        {...ariaFocusableNoLabel(ariaLabel)}
      />
    );
  },
)`
  ${props => iconBaseStyles(props)}
`;

/**
 * General icon-component
 */
export class Icon extends Component<IconProps> {
  render() {
    const { color, icon, tokens, ...passProps } = withSuomifiDefaultProps(
      this.props,
    );
    const { className, ariaLabel } = this.props;

    const iconColor = color !== undefined ? color : tokens.colors.depthDark27;

    if (icon !== undefined) {
      return <StyledSuomifiIcon {...passProps} icon={icon} color={iconColor} />;
    }

    iconLogger(ariaLabel, className);

    return;
  }
}
