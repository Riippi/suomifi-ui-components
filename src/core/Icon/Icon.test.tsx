import React from 'react';
import { render } from 'react-testing-library';
import { axeTest } from '../../utils/test/axe';

import { Icon } from './Icon';

const testIcon = (
  <Icon ariaLabel="Test label" className="my-icon--test" data-testid="icon" />
);

const TestIcons = (
  <div>
    {testIcon}
    <Icon icon="heart" />;
  </div>
);

test('calling render with the same component on the same container does not remount', () => {
  const svgRenderer = render(TestIcons);
  const { getByTestId, container } = svgRenderer;
  expect(container.firstChild).toMatchSnapshot();
  expect(getByTestId('icon')).toBeDefined();

  // re-render the same component with different props
  const testRenderer = render(testIcon);
  const { rerender } = testRenderer;
  rerender(
    <Icon
      ariaLabel="Test label"
      className="my-icon--test"
      data-testid="noci"
    />,
  );
  expect(getByTestId('noci')).toBeDefined();
});

test('should not have basic accessibility issues', axeTest(TestIcons));
