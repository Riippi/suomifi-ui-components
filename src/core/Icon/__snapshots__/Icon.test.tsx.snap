// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`calling render with the same component on the same container does not remount 1`] = `
.c0 {
  display: inline-block;
  vertical-align: baseline;
}

<div>
  <svg
    aria-label="Test label"
    class="my-icon--test c0"
    data-testid="icon"
    height="1em"
    role="img"
    style="fill: hsl(202, 7%, 40%);"
    viewBox="0 0 24 24"
    width="1em"
  >
    <path
      d="M22.01 0c1.103 0 2 .897 2 2v20c0 1.103-.897 2-2 2H10c-1.103 0-2-.897-2-2v-5a1 1 0 012 0v5h12.01V2H10v5a1 1 0 01-2 0V2c0-1.103.897-2 2-2h12.01zm-9.717 15.293L14.586 13H1a1 1 0 010-2h13.586l-2.293-2.293a.999.999 0 111.414-1.414l4 4a.999.999 0 01.216 1.089.999.999 0 01-.216.325l-4 4a.997.997 0 01-1.414 0 .999.999 0 010-1.414z"
    />
  </svg>
  <svg
    aria-hidden="true"
    class="c0"
    focusable="false"
    height="1em"
    style="fill: hsl(202, 7%, 40%);"
    viewBox="0 0 24 24"
    width="1em"
  >
    <path
      d="M17.3 2a6.59 6.59 0 00-4.734 2.001l-.575.588-.574-.588a6.589 6.589 0 00-4.732-2 6.593 6.593 0 00-4.734 2C-.65 6.67-.65 11.034 1.952 13.702l8.608 8.822a1.99 1.99 0 001.431.604c.52 0 1.04-.2 1.432-.604l8.608-8.822c2.603-2.668 2.603-7.033 0-9.7A6.587 6.587 0 0017.3 2M6.685 4a4.58 4.58 0 013.301 1.397l.575.59 1.43 1.466 1.432-1.467.574-.589A4.586 4.586 0 0117.3 4 4.58 4.58 0 0120.6 5.398c1.858 1.904 1.858 5.003 0 6.907l-8.609 8.823-8.607-8.823c-1.858-1.904-1.858-5.003 0-6.907A4.583 4.583 0 016.684 4"
    />
  </svg>
  ;
</div>
`;
